jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -5466,12 +5494,255 @@ public Object helper(Object[] arguments) throws Throwable {
+				}
+			} else {
+				if (relatedMHsList == null) {
+					relatedMHsList = new ArrayList<>();
+				}
+				findRelatedMHs(mh, relatedMHsList);
+			}
+			if ((relatedMHsList == null) || relatedMHsList.isEmpty()) {
+				break;
+			} else {
+				mh = relatedMHsList.remove(relatedMHsList.size() - 1);
+			}
+		}
+		Class<?>[] exceptionTypesArray = null;
+		if (exceptionTypesList != null) {
+			exceptionTypesArray = exceptionTypesList.toArray(new Class<?>[exceptionTypesList.size()]);
+		}
+		return exceptionTypesArray;
+	}
+	
+	/**
+	 * A MethodHandle can be combination of multiple MethodHandles. This helper finds all the related
+	 * MethodHandles for a MethodHandle and appends the related MethodHandles to the relatedMHs list.
+	 * 
+	 * @param mh The MethodHandle to evaluate.
+	 * @param[out] relatedMHs A list of related MethodHandles.
+	 * 
+	 * @return void.
+	 * @throws InternalError for an unknown MethodHandle kind or an exception during a privileged
+	 * operation for ReceiverBoundHandle.
+	 */
+	private static void findRelatedMHs(MethodHandle mh, List<MethodHandle> relatedMHs) {
+		byte kind = mh.kind;
+		if (mh instanceof PrimitiveHandle) {
+			/* Do not append to relatedMHs. */
+		} else if ((kind == MethodHandle.KIND_CONSTANTOBJECT)
+			|| (kind == MethodHandle.KIND_CONSTANTINT)
+			|| (kind == MethodHandle.KIND_CONSTANTFLOAT)
+			|| (kind == MethodHandle.KIND_CONSTANTFLOAT)
+			|| (kind == MethodHandle.KIND_CONSTANTLONG)
+			|| (kind == MethodHandle.KIND_CONSTANTDOUBLE)
+		) {
+			/* Does not support checked exceptions. */
+		} else if (kind == MethodHandle.KIND_VARARGSCOLLECT) {
+			relatedMHs.add(((VarargsCollectorHandle)mh).next);
+		} else if (kind == MethodHandle.KIND_ASTYPE) {
+			relatedMHs.add(((AsTypeHandle)mh).next);
+		} else if (kind == MethodHandle.KIND_FILTERRETURN) {
+			relatedMHs.add(((FilterReturnHandle)mh).next);
+		} else if (kind == MethodHandle.KIND_EXPLICITCAST) {
+			relatedMHs.add(((ExplicitCastHandle)mh).next);
+		} else if (kind == MethodHandle.KIND_SPREAD) {
+			relatedMHs.add(((SpreadHandle)mh).next);
+		} else if (kind == MethodHandle.KIND_FOLDHANDLE) {
+			relatedMHs.add(((FoldHandle)mh).next);
+			relatedMHs.add(((FoldHandle)mh).combiner);
+		} else if (kind == MethodHandle.KIND_GUARDWITHTEST) {
+			relatedMHs.add(((GuardWithTestHandle)mh).guard);
+			relatedMHs.add(((GuardWithTestHandle)mh).trueTarget);
+			relatedMHs.add(((GuardWithTestHandle)mh).falseTarget);
+		} else if (kind == MethodHandle.KIND_FILTERARGUMENTS) {
+			relatedMHs.add(((FilterArgumentsHandle)mh).next);
+			Collections.addAll(relatedMHs, ((FilterArgumentsHandle)mh).filters);
+		} else if (kind == MethodHandle.KIND_FILTERARGUMENTS_WITHCOMBINER) {
+			relatedMHs.add(((FilterArgumentsWithCombinerHandle)mh).next);
+			relatedMHs.add(((FilterArgumentsWithCombinerHandle)mh).combiner);
+		} else if (kind == MethodHandle.KIND_DYNAMICINVOKER) {
+			relatedMHs.add(((DynamicInvokerHandle)mh).site.getTarget());
```

Comment:
Is this the expected behaviour?  I think you can create an infinite loop here by doing something like:
```java
MutableCallSite mcs = new MutableCallSite(MethodType.type(void.class));
MH dynInvoker = mcs.dynamicInvoker();
mcs.setTarget(dynInvoker);
checkedExceptions.dynInvoker();
```

Reply:
@DanHeidinga No, we want to avoid such cycles (infinite loops). If the parent and next MHs are the same, then the next MH should be skipped. Is this scenario only possible for `DynamicInvokerHandle`?

Reply:
DynamicInvokerHandle is the only case where I know this can occur as other other handles are immutable once built so they can't refer to themselves.

The question of "expected behaviour" was whether checkedExceptions should be found in the target MH chain or not.  I would have expected the chain to the dynamicInvoker to be considered "complete" and the target MH chain to be separate, at least for the case of MutableCallSite or VolatileCallSite

Reply:
Ok. Shall I use `CallSite.dynamicInvoker()` instead of `CallSite.getTarget()` to retrieve the `DynamicInvokerHandle`'s target MH since it's "complete"?